src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs

```diff
@@ -181,6 +188,11 @@ protected virtual void InitializeTransactionBuilder(TransactionBuildContext cont
             Guard.NotNull(context.Recipients, nameof(context.Recipients));
             Guard.NotNull(context.AccountReference, nameof(context.AccountReference));
 
+            context.TransactionBuilder.CoinSelector = new DefaultCoinSelector
+            {
+                GroupByScriptPubKey = false
```

Comment:
@MithrilMan can you explain again what the reason was behind this change?

Reply:
Actually this is quite controversial:
GroupByScriptPubKey  is a setting used to group unspent coins by ScriptPubKey (for P2PKH you can read this as "by address")

when this setting is true, transactionbuilder tries to find the better set of those utxos to be spent, and spend them all.
What this mean is that, suppose you have 10 utxo so composed:
(for simplicity i use letters to indicate different ScriptPubKeys)

Amount | ScriptPubKey
---|---
1|A
3|A
5|A
5|B
10|B
1|B
3|B
10|C
2|C
2|C

Now suppose that you ask TB to build a transaction that spend 2 coin and that TB picks up the 2nd group as the chosen to be spent, this mean that our inputs will spend these coins

Amount | ScriptPubKey
---|---
5|B
10|B
1|B
3|B

even if the amount of the transaction could be covered by a single coin (the one of value 3) all these inputs will be used instead.

The reason is to protect privacy, because you are burning all coins that belongs to the same "address" and don't mix it up with other addresses that some transaction analyzer may link with other movements on your wallet.

At the same time this has the pros of burning more utxos than what's it's generating (because as outputs we have the destination one with value of 2 and a change) shrinking the UTXO set that's always considered a good thing for blockchain scalability because UTXO set has to fit in RAM and the less UTXO you have to store, the better.

But there is a con: the fee!! Yes because now we are generating a transaction that uses 4 inputs instead of one and since we pay a fee based on the transaction size, we spend way more.

The change I did was actually to disable such flag, sacrificing privacies for tx size. The reason tbh was even more practical: doing so I didn't had to fix tests that were expecting a specified amount of fee, because of course the fee was higher then expected.

For the same reason I added the code that throw away the non useful UTXO at the end of the transaction that has been built.

We may still consider to revert this back or come out with our own version of transactionbuilder if needed in the future. (and custom coin selectors)

Reply:
maybe @NicolasDorier could add some additional considerations about this

Reply:
Well personally I don't think it is a good thing to disable it because of privacy concerns as explained. 

If you do that mainly for making the tests happy, you can instead put a property in the WalletTransactionHandler to opt in for disabling grouping and set this boolean in your tests.